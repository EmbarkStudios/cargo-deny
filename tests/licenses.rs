use cargo_deny::{diag, field_eq, func_name, licenses, test_utils as tu, Krates};
use parking_lot::Once;
use std::sync::Arc;

static mut STORE: Option<Arc<licenses::LicenseStore>> = None;
static INIT: Once = Once::new();

fn store() -> Arc<licenses::LicenseStore> {
    #[allow(unsafe_code)]
    unsafe {
        INIT.call_once(|| {
            STORE = Some(Arc::new(licenses::LicenseStore::from_cache().unwrap()));
        });
        STORE.as_ref().unwrap().clone()
    }
}

#[inline]
pub fn gather_licenses_with_overrides(
    name: &str,
    cfg: impl Into<tu::Config<crate::licenses::cfg::Config>>,
    overrides: Option<diag::DiagnosticOverrides>,
) -> Vec<serde_json::Value> {
    let md: krates::cm::Metadata = serde_json::from_str(
        &std::fs::read_to_string("tests/test_data/features-galore/metadata.json").unwrap(),
    )
    .unwrap();

    let krates: Krates = krates::Builder::new()
        .build_with_metadata(md, krates::NoneFilter)
        .unwrap();

    let gatherer = licenses::Gatherer::default()
        .with_store(store())
        .with_confidence_threshold(0.8);

    let mut files = codespan::Files::new();

    let cfg = cfg.into();
    let lic_cfg = {
        let des: licenses::cfg::Config = toml::from_str(&cfg.config).unwrap();
        let cfg_id = files.add("config.toml", cfg.config.clone());

        let mut diags = Vec::new();
        use cargo_deny::UnvalidatedConfig;
        des.validate(cfg_id, &mut diags)
    };

    let summary = gatherer.gather(&krates, &mut files, Some(&lic_cfg));

    tu::gather_diagnostics_with_files::<crate::licenses::cfg::Config, _, _>(
        &krates,
        name,
        cfg,
        files,
        |ctx, _cs, tx| {
            crate::licenses::check(
                ctx,
                summary,
                diag::ErrorSink {
                    overrides: overrides.map(Arc::new),
                    channel: tx,
                },
            );
        },
    )
}

#[test]
fn accepts_licenses() {
    let cfg = tu::Config::new(
        "allow = ['Apache-2.0', 'MIT']
    exceptions = [{ name = 'tinyvec_macros', allow = ['Zlib']}]",
    );

    let diags = gather_licenses_with_overrides(func_name!(), cfg, None);

    insta::assert_json_snapshot!(diags);
}

#[test]
fn rejects_licenses() {
    let cfg = tu::Config::new("allow = []");

    let diags = gather_licenses_with_overrides(func_name!(), cfg, None);

    insta::assert_json_snapshot!(diags);
}

#[test]
fn accepts_exceptions() {
    let cfg = tu::Config::new("exceptions = [{ name = 'tinyvec_macros', allow = ['Zlib']}]");

    let mut diags = gather_licenses_with_overrides(func_name!(), cfg, None);

    // Just keep tinyvec and tinyvec_macros which both have the same licenses,
    // but since we only allowed the exception on macros, that is the only
    // one that should succeed
    diags.retain(|d| {
        field_eq!(d, "/fields/graphs/0/Krate/name", "tinyvec")
            || field_eq!(d, "/fields/graphs/0/Krate/name", "tinyvec_macros")
    });

    insta::assert_json_snapshot!(diags);
}

#[test]
fn detects_unlicensed() {
    let cfg = tu::Config::new("unlicensed = 'warn'");

    let mut diags = gather_licenses_with_overrides(func_name!(), cfg, None);

    diags.retain(|d| field_eq!(d, "/fields/severity", "warning"));

    insta::assert_json_snapshot!(diags);
}

#[test]
fn flags_unencountered_licenses() {
    let cfg = tu::Config::new(
        "allow = ['Aladdin', 'MIT']
    unlicensed = 'allow'",
    );

    // Override the warning to be a failure
    let overrides = cargo_deny::overrides! {
        "license-not-encountered" => Error,
    };

    let mut diags = gather_licenses_with_overrides(func_name!(), cfg, Some(overrides));

    diags.retain(|d| field_eq!(d, "/fields/severity", "error"));

    insta::assert_json_snapshot!(diags);
}

#[test]
fn flags_unencountered_exceptions() {
    let cfg = tu::Config::new(
        "allow = ['MIT']
    unlicensed = 'allow'
    exceptions = [{name='bippity-boppity-boop', allow = ['Aladdin']}]",
    );

    // Override the warning to be a failure
    let overrides = cargo_deny::overrides! {
        "license-exception-not-encountered" => Error,
    };

    let mut diags = gather_licenses_with_overrides(func_name!(), cfg, Some(overrides));

    diags.retain(|d| field_eq!(d, "/fields/severity", "error"));

    insta::assert_json_snapshot!(diags);
}

/// Ensures that invalid SPDX expressions in strict mode can be parsed when
/// falling back to more lax rules, but still output a warning
#[test]
fn lax_fallback() {
    let mut cmd = krates::Cmd::new();
    cmd.manifest_path("examples/04_gnu_licenses/Cargo.toml");

    let krates: Krates = krates::Builder::new()
        .build(cmd, krates::NoneFilter)
        .unwrap();

    let gatherer = licenses::Gatherer::default()
        .with_store(store())
        .with_confidence_threshold(0.8);

    let mut files = codespan::Files::new();

    let cfg: tu::Config<crate::licenses::cfg::Config> = tu::Config::new(
        "allow = ['GPL-2.0', 'LGPL-3.0']
    unlicensed = 'deny'",
    );

    let lic_cfg = {
        let des: licenses::cfg::Config = toml::from_str(&cfg.config).unwrap();
        let cfg_id = files.add("config.toml", cfg.config.clone());

        let mut diags = Vec::new();
        use cargo_deny::UnvalidatedConfig;
        des.validate(cfg_id, &mut diags)
    };

    let summary = gatherer.gather(&krates, &mut files, Some(&lic_cfg));

    let diags = tu::gather_diagnostics_with_files::<crate::licenses::cfg::Config, _, _>(
        &krates,
        "lax_fallback",
        cfg,
        files,
        |ctx, _cs, tx| {
            crate::licenses::check(
                ctx,
                summary,
                diag::ErrorSink {
                    overrides: None,
                    channel: tx,
                },
            );
        },
    );

    insta::assert_json_snapshot!(diags);
}
