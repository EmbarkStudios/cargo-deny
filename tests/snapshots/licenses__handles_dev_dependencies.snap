---
source: tests/licenses.rs
expression: diags
---
[
  {
    "fields": {
      "code": "rejected",
      "graphs": [
        {
          "Krate": {
            "name": "dynamic",
            "version": "0.2.1"
          },
          "parents": [
            {
              "Krate": {
                "name": "simple_ecs",
                "version": "0.3.2"
              },
              "parents": [
                {
                  "Krate": {
                    "kind": "dev",
                    "name": "features-galore",
                    "version": "0.1.0"
                  }
                }
              ]
            }
          ]
        }
      ],
      "labels": [
        {
          "column": 12,
          "line": 4,
          "message": "license expression retrieved via Cargo.toml `license`",
          "span": "MIT"
        },
        {
          "column": 12,
          "line": 4,
          "message": "rejected: license is not explicitly allowed",
          "span": "MIT"
        }
      ],
      "message": "failed to satisfy license requirements",
      "notes": [
        "MIT - MIT License:",
        "  - OSI approved",
        "  - FSF Free/Libre"
      ],
      "severity": "error"
    },
    "type": "diagnostic"
  },
  {
    "fields": {
      "code": "rejected",
      "graphs": [
        {
          "Krate": {
            "name": "simple_ecs",
            "version": "0.3.2"
          },
          "parents": [
            {
              "Krate": {
                "kind": "dev",
                "name": "features-galore",
                "version": "0.1.0"
              }
            }
          ]
        }
      ],
      "labels": [
        {
          "column": 12,
          "line": 4,
          "message": "a deprecated license identifier was used",
          "span": "GPL-3.0"
        },
        {
          "column": 12,
          "line": 4,
          "message": "license expression retrieved via Cargo.toml `license`",
          "span": "GPL-3.0"
        },
        {
          "column": 12,
          "line": 4,
          "message": "rejected: license is not explicitly allowed",
          "span": "GPL-3.0"
        }
      ],
      "message": "failed to satisfy license requirements",
      "notes": [
        "GPL-3.0 - GNU General Public License v3.0 only:",
        "  - **DEPRECATED**",
        "  - OSI approved",
        "  - FSF Free/Libre",
        "  - Copyleft"
      ],
      "severity": "error"
    },
    "type": "diagnostic"
  }
]
